### PR TITLE
Extended export plugin support

### PR DESCRIPTION
**Pull request by austinmm**
The original export plugin only exported as JSON so I added the ability to export in CSV or XML format. To do this I created a new option `-f` or `--format` which specifies the export format, the default is still JSON. For example...
`$ beet export -f csv -i "title,album" ACDC`
I also updated the plugin documentation to reflect the changes I've made to the plugin. Lastly, I also created a unit test for the export plugin; since one hadn't been created yet.

### CODE REVIEW

**Comment by sampsyo** [comment]
Cool; thanks for getting this started!

Here's one question for context: while I can totally see the utility of a CSV export, I'm struggling to think of a use case for an XML export. Can you elaborate a little more on what you want to use CSV and XML data for? I think that would help calibrate our approach to these features.

File: beetsplug/export.py [hunk 1]
```diff
@@ -40,6 +42,9 @@ class ExportPlugin(BeetsPlugin):
 
     def __init__(self):
         super(ExportPlugin, self).__init__()
+        # Used when testing export plugin
+        self.run_results = None
```
Comment (sampsyo) [inline beetsplug/export.py]: Hmm; I don't think I see where this is actually used?

File: beetsplug/export.py [hunk 2]
```diff
@@ -40,6 +42,9 @@ class ExportPlugin(BeetsPlugin):
...
+        self.export_format = None
```
Comment (sampsyo) [inline beetsplug/export.py]: I also don't quite see why this is a class field instead of just being a local variable in the `run` method. Can we please revert to the old way?

File: beetsplug/export.py [hunk 3]
```diff
@@ -52,6 +57,24 @@ def __init__(self):
...
+                }
+            },
+            'xml': {
+                # xml module formatting options
+                'formatting': {
+                    'ensure_ascii': False,
+                    'indent': 4,
+                    'separators': ('>'),
+                    'sort_keys': True
+                }
```
Comment (sampsyo) [inline beetsplug/export.py]: I don't think any of these config options are actually used in the CSV or XML emitters.

File: beetsplug/export.py [hunk 4]
```diff
@@ -78,17 +101,21 @@ def commands(self):
             u'-o', u'--output',
             help=u'path for the output file. If not given, will print the data'
         )
+        cmd.parser.add_option(
+            u'-f', u'--format', default='json',
+            help=u'specify the format of the exported data. Your options are json (deafult), csv, and xml'
```
Comment (sampsyo) [inline beetsplug/export.py]: This could be shorter: perhaps, "the output format: json (default), csv, or xml".

File: beetsplug/export.py [hunk 5]
```diff
@@ -78,17 +101,21 @@ def commands(self):
...
+        )
         return [cmd]
 
     def run(self, lib, opts, args):
-
         file_path = opts.output
-        file_format = self.config['default_format'].get(str)
         file_mode = 'a' if opts.append else 'w'
+        file_format = opts.format
```
Comment (sampsyo) [inline beetsplug/export.py]: Please preserve the effect of `default_format`.

File: beetsplug/export.py [hunk 6]
```diff
@@ -108,44 +135,118 @@ def run(self, lib, opts, args):
             except (mediafile.UnreadableFileError, IOError) as ex:
                 self._log.error(u'cannot read file: {0}', ex)
                 continue
-
```
Comment (sampsyo) [inline beetsplug/export.py]: Any particular reason to delete this newline?

File: beetsplug/export.py [hunk 7]
```diff
@@ -108,44 +135,118 @@ def run(self, lib, opts, args):
...
             data = key_filter(data)
             items += [data]
 
-        export_format.export(items, **format_options)
+        self.run_results = items
+        self.export_format.export(self.run_results, **format_options)
```
Comment (sampsyo) [inline beetsplug/export.py]: As mentioned above, I think it would be awesome to remove these side effects on mutable class fields…

File: beetsplug/export.py [hunk 8]
```diff
@@ -108,44 +135,118 @@ def run(self, lib, opts, args):
...
-    def factory(cls, type, **kwargs):
-        if type == "json":
-            if kwargs['file_path']:
-                return JsonFileFormat(**kwargs)
-            else:
-                return JsonPrintFormat()
-        raise NotImplementedError()
+    def factory(cls, file_type, **kwargs):
+        if file_type == "json":
+            return JsonFormat(**kwargs)
```
Comment (sampsyo) [inline beetsplug/export.py]: I notice this now behaves differently from it did before—the `kwargs` are now forwarded unconditionally instead of only when the path is unspecified. Does this change the user-visible behavior?

File: beetsplug/export.py [hunk 9]
```diff
@@ -108,44 +135,118 @@ def run(self, lib, opts, args):
...
         raise NotImplementedError()
 
 
-class JsonPrintFormat(ExportFormat):
-    """Outputs to the console"""
+class JsonFormat(ExportFormat):
+    """Saves in a json file"""
+    def __init__(self, file_path, file_mode=u'w', encoding=u'utf-8'):
+        super(JsonFormat, self).__init__(file_path, file_mode, encoding)
+        self.export = self.export_to_file if self.path else self.export_to_terminal
```
Comment (sampsyo) [inline beetsplug/export.py]: This technique where a method is defined conditionally is somewhat surprising. I now see how it works after some puzzling about it, but it could use an explanatory comment.

File: beetsplug/export.py [hunk 10]
```diff
@@ -108,44 +135,118 @@ def run(self, lib, opts, args):
...
+        writer = csv.DictWriter(sys.stdout, fieldnames=self.header)
+        writer.writeheader()
+        writer.writerows(data)
+        self.results = str(writer)
 
-class JsonFileFormat(ExportFormat):
-    """Saves in a json file"""
 
+    def export_to_file(self, data, **kwargs):
+        with codecs.open(self.path, self.mode, self.encoding) as f:
```
Comment (sampsyo) [inline beetsplug/export.py]: I think it would be really useful to factor out the file-opening part of the logic so it doesn't need to be re-implemented for every export format. In general, the output format and output destination (file or stdout) seem orthogonal.

File: beetsplug/export.py [hunk 11]
```diff
@@ -108,44 +135,118 @@ def run(self, lib, opts, args):
...
+
+class CSVFormat(ExportFormat):
+    """Saves in a csv file"""
+    def __init__(self, file_path, file_mode=u'w', encoding=u'utf-8'):
+        super(CSVFormat, self).__init__(file_path, file_mode, encoding)
+        self.header = []
 
     def export(self, data, **kwargs):
-        json.dump(data, sys.stdout, cls=ExportEncoder, **kwargs)
+        if data and type(data) is list and len(data) > 0:
```
Comment (sampsyo) [inline beetsplug/export.py]: I'm a little confused by the `list` type check here—when will the passed object _not_ be a list? And what happens when it's not—so `self.header` remains empty?

Also, this condition could be simplified to:

```py
if data and isinstance(data, list):
```

because `data` is only "true" if it is non-empty.

File: docs/plugins/export.rst [hunk 12]
```diff
@@ -36,11 +39,18 @@ The ``export`` command has these command-line options:
 
 * ``--append``: Appends the data to the file instead of writing.
 
+* ``--format`` or ``-f``: Specifies the format the data will be exported as. If not informed, JSON will be used by default.
+  For example::
+
+      $ beet export -f csv beatles
+      $ beet export -f json beatles
+      $ beet export -f xml beatles
```
Comment (sampsyo) [inline docs/plugins/export.rst]: I think we can do without the full command-line examples here. Can we just list the options?

File: docs/plugins/export.rst [hunk 13]
```diff
@@ -63,3 +73,15 @@ The default options look like this::
...
+                ensure_ascii: False
+                indent: 0
+                separators: [',']
+                sort_keys: true
+        xml:
+            formatting:
+                ensure_ascii: False
+                indent: 4
+                separators: ['>']
+                sort_keys: true
```
Comment (sampsyo) [inline docs/plugins/export.rst]: As above, I don't think these options are actually used for the new formats.

File: beetsplug/export.py (hunk 1 above)
Comment (austinmm) [reply beetsplug/export.py]: I was going to use that class field as a way of checking the output during the testing of the export plugin but I ended up going another route so I will go ahead and remove it.

File: beetsplug/export.py (hunk 2 above)
Comment (austinmm) [reply beetsplug/export.py]: I reverted it.

File: beetsplug/export.py (hunk 4 above)
Comment (austinmm) [reply beetsplug/export.py]: Sounds good

File: beetsplug/export.py (hunk 5 above)
Comment (austinmm) [reply beetsplug/export.py]: would you prefer I get rid of the `default=json` in...
```
cmd.parser.add_option(
            u'-f', u'--format', default='json',
            help=u"the output format: json (default), csv, or xml"
        )
```
and instead have something like...
```
file_format = opts.format if opts.format else self.config['default_format'].get(str)
```

File: beetsplug/export.py (hunk 9 above)
Comment (austinmm) [reply beetsplug/export.py]: Sounds good, I'll add some comments explaining its purpose

File: beetsplug/export.py (hunk 11 above)
Comment (austinmm) [reply beetsplug/export.py]: I was just trying to be extra cautious but I completely agree that it is a little overkill, this should work just fine instead; `if data and len(data) > 0:`

File: docs/plugins/export.rst (hunk 12 above)
Comment (austinmm) [reply docs/plugins/export.rst]: Sounds good

File: beetsplug/export.py (hunk 3 above)
Comment (austinmm) [reply beetsplug/export.py]: I was just trying to follow the same format, but you are right they are not being used. I have updated them to now be format specific to json, csv and xml and updated the csv and xml outputs to use them...
```
'json': {
                # json module formatting options
                'formatting': {
                    'ensure_ascii': False,
                    'indent': 4,
                    'separators': (',', ': '),
                    'sort_keys': True
                }
            },
            'csv': {
                # csv module formatting options
                'formatting': {
                    'delimiter': ',', # column seperator
                    'dialect': 'excel', # the name of the dialect to use
                    'quotechar':'|'
                }
            },
            'xml': {
                # xml module formatting options
                'formatting': {
                    'encoding': 'UTF-8', # the output encoding
                    'xml_declaration':'True', # controls if an XML declaration should be added to the file
                    'default_namespace': 'None', # sets the default XML namespace (for “xmlns”)
                    'method': 'xml', # either "xml", "html" or "text" (default is "xml")
                    'short_empty_elements': 'True' # controls the formatting of elements that contain no content.
                }
```

File: beetsplug/export.py (hunk 6 above)
Comment (austinmm) [reply beetsplug/export.py]: Nope did that by mistake, I added it back

File: beetsplug/export.py (hunk 7 above)
Comment (austinmm) [reply beetsplug/export.py]: will do

File: docs/plugins/export.rst (hunk 13 above)
Comment (austinmm) [reply docs/plugins/export.rst]: I updated them to work correctly now with the added formats CSV and XML

**Comment by austinmm** [comment]
I am still working on all the changes but almost done. Thank you for all the feedback as it was very helpful and constructive. Also to address your question about the XML format. The reason I added the ability to export in the XML format was that this is the preferred export format used by Apple in Apple Music and simply figured that if Apple does it then there must exist a good reason. For CSV exporting, I imagined that some people, including myself, might want to use Excel tools to better analyze their music library.

File: beetsplug/export.py (hunk 8 above)
Comment (austinmm) [reply beetsplug/export.py]: This should not effect the user in any way. It simply simplifies the logic in my opinion.

File: beetsplug/export.py (hunk 10 above)
Comment (austinmm) [reply beetsplug/export.py]: I'll explore that idea. Sounds like a better way

**Comment by austinmm** [comment]
I have just pushed the changes that were discussed above. Please let me know if you recommend any other changes. :)

File: beetsplug/export.py [hunk 14]
```diff
@@ -52,6 +54,23 @@ def __init__(self):
                     'sort_keys': True
                 }
             },
+            'csv': {
+                # csv module formatting options
+                'formatting': {
+                    'delimiter': ',', # column seperator
+                    'dialect': 'excel', # the name of the dialect to use
+                    'quotechar': '|'
```
Comment (sampsyo) [inline beetsplug/export.py]: I think you probably want `quotechar` to default to "? In fact, it doesn't seem like the most useful thing to want to override… maybe we can get away without this option.

File: beetsplug/export.py [hunk 15]
```diff
@@ -52,6 +54,23 @@ def __init__(self):
                     'sort_keys': True
                 }
             },
+            'csv': {
+                # csv module formatting options
+                'formatting': {
+                    'delimiter': ',', # column seperator
+                    'dialect': 'excel', # the name of the dialect to use
```
Comment (sampsyo) [inline beetsplug/export.py]: This option seems useful enough! But maybe it's worth linking (in the documentation) to the place in the Python docs where dialects are described?

File: beetsplug/export.py [hunk 16]
```diff
@@ -117,35 +140,74 @@ def run(self, lib, opts, args):
...
-class JsonFileFormat(ExportFormat):
-    """Saves in a json file"""
+class CSVFormat(ExportFormat):
+    """Saves in a csv file"""
+    def __init__(self, file_path, file_mode=u'w', encoding=u'utf-8'):
+        super(CSVFormat, self).__init__(file_path, file_mode, encoding)
+        self.header = []
 
+    def export(self, data, **kwargs):
+        if data and len(data) > 0:
```
Comment (sampsyo) [inline beetsplug/export.py]:
```suggestion
        if data:
```

File: docs/plugins/export.rst [hunk 17]
```diff
@@ -63,3 +69,15 @@ The default options look like this::
...
```
Comment (sampsyo) [inline docs/plugins/export.rst]: I believe these examples are now out of sync with the implementation. It would also be good to describe the options in English above.

File: beetsplug/export.py (hunk 14 above)
Comment (austinmm) [reply beetsplug/export.py]: Sounds good, I'll remove it

File: beetsplug/export.py (hunk 15 above)
Comment (austinmm) [reply beetsplug/export.py]: Oh yes I forgot to do that, thank you for reminding me

File: beetsplug/export.py (hunk 16 above)
Comment (austinmm) [reply beetsplug/export.py]: Will do

**Comment by austinmm** [comment]
I have made the changes you suggested alongside some other just general code cleanup. I am not sure where I might be disobeying the PEP-8 conventions so I apologize if I am anywhere. I am still new to CIs and so would love any recommendations you might have so I can pass the two CI test.

**Comment by sampsyo** [comment]
For the CI results, take a look at GitHub's "checks" interface on this page. It has "Details" links to the logs from Travis and CircleCI. For example, here's the log from the style checker:
https://travis-ci.org/beetbox/beets/jobs/597112966

You can also consider installing [flake8](http://flake8.pycqa.org/en/latest/) and running it locally.

Also, about the docs: while the text seems good, could you please move these from inline comments in the example YAML to the prose text above? There's already a section there with a bulleted list describing the options, before the page gets to the YAML example. That would be a nice place to add similar descriptions of the new options.

**Comment by austinmm** [comment]
Both the test_export.py and export.py pass the flake8 test so I'm not sure why there is still an issue with the CLI. How do I run the test_export.py on my computer because if I just run `python test/test_export.py` it doesn't compile correctly and neither do any of the other python tests files?

**Comment by sampsyo** [comment]
We have [a wiki page about testing](https://github.com/beetbox/beets/wiki/Testing) that might help! You might want to try using Tox or check out the section about test dependencies.

You can also see the results of the tests directly on Travis:
https://travis-ci.org/beetbox/beets/jobs/597334455

**Comment by austinmm** [comment]
I was able to get all the tests to pass :). Please let me know if you want me to update anything else, but I believe it is production-ready.

File: beetsplug/export.py [hunk 18]
```diff
@@ -78,17 +100,22 @@ def commands(self):
...
+        )
         return [cmd]
 
     def run(self, lib, opts, args):
-
         file_path = opts.output
-        file_format = self.config['default_format'].get(str)
         file_mode = 'a' if opts.append else 'w'
+        file_format = opts.format if opts.format else \
+            self.config['default_format'].get(str)
```
Comment (sampsyo) [inline beetsplug/export.py]: This should suffice: `opts.format or self.config['default_format'].get(str)`

File: beetsplug/export.py [hunk 19]
```diff
@@ -117,35 +145,79 @@ def run(self, lib, opts, args):
...
 class ExportFormat(object):
     """The output format type"""
+    def __init__(self, file_path, file_mode=u'w', encoding=u'utf-8'):
+        self.path = file_path
+        self.mode = file_mode
+        self.encoding = encoding
+        """ self.out_stream =
+                sys.stdout if path doesn't exit
+                codecs.open(..) else
+        """
```
Comment (sampsyo) [inline beetsplug/export.py]: Perhaps this is out of date? Or it can be rephrased as a natural-language comment?

File: beetsplug/export.py [hunk 20]
```diff
@@ -117,35 +145,79 @@ def run(self, lib, opts, args):
...
+                track_key = ET.SubElement(tracks_dict, u'key')
+                track_key.text = str(index)
+                track_dict = ET.SubElement(tracks_dict, u'dict')
+                track_details = ET.SubElement(track_dict, u'Track ID')
+                track_details.text = str(index)
+                for key, value in item.items():
+                    track_details = ET.SubElement(track_dict, key)
+                    track_details.text = value
+
+        # tree = ET.ElementTree(element=library)
```
Comment (sampsyo) [inline beetsplug/export.py]: Out of date?

File: docs/plugins/export.rst [hunk 21]
```diff
@@ -36,19 +39,34 @@ The ``export`` command has these command-line options:
...
 
-- **ensure_ascii**: Escape non-ASCII characters with ``\uXXXX`` entities.
+    - **sort_keys**: Sorts the keys in JSON dictionaries.
 
-- **indent**: The number of spaces for indentation.
+- **CSV Formatting**
+    - **delimiter**: Used as the separating character between fields. The default value is a comma (,).
 
-- **separators**: A ``[item_separator, dict_separator]`` tuple.
+    - **dialect**: A dialect, in the context of reading and writing CSVs, is a construct that allows you to create, store, and re-use various formatting parameters for your data.
```
Comment (sampsyo) [inline docs/plugins/export.rst]: Instead of duplicating the text from the Python CSV documentation, can we instead link to the explanation there?

File: docs/plugins/export.rst [hunk 22]
```diff
@@ -36,19 +39,34 @@ The ``export`` command has these command-line options:
...
 
-- **sort_keys**: Sorts the keys in JSON dictionaries.
+- **XML Formatting**
+    - **encoding**: Use encoding="unicode" to generate a Unicode string (otherwise, a bytestring is generated).
```
Comment (sampsyo) [inline docs/plugins/export.rst]: Hmm… how does this affect the usage of the plugin? Because we're writing to a file, it seems like the kind of Python string that gets generated is irrelevant. Perhaps there's an argument to be made that this should _not_ be configurable in that case?

File: docs/plugins/export.rst [hunk 23]
```diff
@@ -36,19 +39,34 @@ The ``export`` command has these command-line options:
...
+
+    - **xml_declaration**: Controls if an XML declaration should be added to the file. Use False for never, True for always, None for only if not US-ASCII or UTF-8 or Unicode (default is None).
```
Comment (sampsyo) [inline docs/plugins/export.rst]: TBH I'm not sure if I know what an "XML declaration" is. Is this something that people will actually want to control?

File: docs/plugins/export.rst [hunk 24]
```diff
@@ -36,19 +39,34 @@ The ``export`` command has these command-line options:
...
+
+    - **method**: Can be either "xml", "html" or "text" (default is "xml")
```
Comment (sampsyo) [inline docs/plugins/export.rst]: Maybe it's worth explaining what these options actually do? I'm not sure how setting `method` to `text` will affect the output, for example.

File: beetsplug/export.py (hunk 19 above)
Comment (austinmm) [reply beetsplug/export.py]: Here is the new comment...
`# creates a file object to write/append or sets to stdout`

File: beetsplug/export.py (hunk 18 above)
Comment (austinmm) [reply beetsplug/export.py]: sounds good

File: beetsplug/export.py (hunk 20 above)
Comment (austinmm) [reply beetsplug/export.py]: yeah

File: docs/plugins/export.rst (hunk 21 above)
Comment (austinmm) [reply docs/plugins/export.rst]: I will link to an external description but I would imagine it is still good to have some description in the document itself. I just shortened it a little. Let me know if you want me to get rid of the description altogether.

File: docs/plugins/export.rst (hunk 22 above)
Comment (austinmm) [reply docs/plugins/export.rst]: I agree, it is not an config option now.

File: docs/plugins/export.rst (hunk 23 above)
Comment (austinmm) [reply docs/plugins/export.rst]: Its been removed

**Comment by austinmm** [comment]
I believe I have successfully implemented all your recommended changes. Let me know if I need to make any other changes. Also, here is the updated XML format...
```
<library>
    <tracks>
        <track>
            <title>Catch Me</title>
            <album>BRÅVES</album>
            ....
        </track>
        ....
    </tracks>
</library>
```

File: docs/plugins/export.rst (hunk 24 above)
Comment (sampsyo) [reply docs/plugins/export.rst]: Hmm… correct me if I'm wrong, but isn't the HTML example here identical to the XML example? Is there a difference I'm missing?

**Comment by sampsyo** [comment]
Cool! Thanks for continuing to make progress!

I have only two more things in mind:

- We'll need a changelog entry about the new feature.
- I'm still uncertain about the utility of XML export formats. It seems to me that "HTML" export is exactly the same as the "XML" format of XML, isn't it? And the text format doesn't seem all that useful to me…

File: docs/plugins/export.rst (hunk 24 above)
Comment (austinmm) [reply docs/plugins/export.rst]: In this case, the HTML is the same but I don't think that is always the case. I agree the text is not very useful. Would you like me to remove this option?

Comment (sampsyo) [reply docs/plugins/export.rst]: I think so! Unless you can think of a compelling reason why someone would want to use it, let's take it out.

**Comment by austinmm** [comment]
How do we create a changelog entry?

**Comment by sampsyo** [comment]
There's a file in `docs/changelog.rst` that serves that purpose. :smiley:

**Comment by austinmm** [comment]
Let me know if I didn't create the changelog entry correctly or if you notice anything else that needs fixing :)

File: docs/plugins/export.rst [hunk 25]
```diff
@@ -36,24 +39,36 @@ The ``export`` command has these command-line options:
...
+    - **sort_keys**: Sorts the keys in JSON dictionaries.
 
 These options match the options from the `Python json module`_.
 
 .. _Python json module: https://docs.python.org/2/library/json.html#basic-usage
 
+- **CSV Formatting**
+    - **delimiter**: Used as the separating character between fields. The default value is a comma (,).
+
+    - **dialect**: A dialect is a construct that allows you to create, store, and re-use various formatting parameters for your data.
```
Comment (sampsyo) [inline docs/plugins/export.rst]:
```suggestion
    - **dialect**: The kind of CSV file to produce. The default is `excel`.
```

Keeping this focused on what matters to *users of this plugin*, not to Python programmers.

File: docs/plugins/export.rst [hunk 26]
```diff
@@ -62,4 +77,8 @@ The default options look like this::
                 ensure_ascii: False
                 indent: 4
                 separators: [',' , ': ']
-                sort_keys: true
+                sort_keys: True
```
Comment (sampsyo) [inline docs/plugins/export.rst]: Lower case is OK here—it’s YAML, not Python.
```suggestion
                sort_keys: true
```

File: docs/plugins/export.rst [hunk 27]
```diff
@@ -62,4 +77,8 @@ The default options look like this::
...
+        csv:
+            formatting:
+                delimiter: ','
+                dialect: 'excel'
```
Comment (sampsyo) [inline docs/plugins/export.rst]:
```suggestion
                dialect: excel
```

No quotes are usually necessary in YAML.

**Comment by sampsyo** [comment]
Arg, turns out GitHub won’t let me commit my own suggestions on my phone. :( But I will wrap this up when I’m back at a normal computer.

**Comment by austinmm** [comment]
Your changes look great! Thank you for all the help you provided me with this. This was my first time contributing to an open-source project so all your assistance was very much appreciated.

**Comment by sampsyo** [comment]
Wow! In that case, thank you for your diligent efforts, congratulations, and welcome to the world of open source. :smiley: Nicely done!
